app/features/flex.js

```diff
@@ -41,16 +42,26 @@ export function Flex({selection}) {
       changeDirection(selectedNodes, 'column')
   })
 
+  visbug.onSelectedUpdate(onSelectedUpdateHandler)
+
   return () => {
     hotkeys.unbind(key_events)
     hotkeys.unbind(command_events)
     hotkeys.unbind('up,down,left,right')
+    visbug.removeSelectedCallback(onSelectedUpdateHandler)
   }
 }
 
-const ensureFlex = el => {
-  el.style.display = 'flex'
-  return el
+export const ensureFlex = (el) => el.style.display = 'flex'
+export const assignLabel = (el) => $(el).attr('data-label', `display: ${getComputedStyle(el).display}`)
+
+const ensureFlexAndAssignLabel = callback => (els, ...args) => {
```

Comment:
i'm in the middle on this! i like your use of a HoF here, but i'm thinking out the tradeoffs:

gained:
- perf (can you elaborate on why you believe this is faster, i have hunches but i dont wanna guess)
- tiny bit more succinct (debatable)

lost:
- when the functions are separate they can move around easier (debatable)
- would/should update all other functional chains to use the same pattern (2 patterns instead of 1 until refactor)
- was more straightforward before than with the abstraction (debatable)

what i like most is how much of the code you went and understood before making decisions. you're totally right in seeing lots of room for refactor improvement, i'm glad you saw those areas, small and large. 

ultimately i'm leaning towards following the chain pattern already there (no higher order functions), but could be swayed. i'm enjoying seeing your code and how you want to express!

thanks for this PR!

Reply:
so yeah, my main concern was performance so let's talk about it first. 
### Pros
#### Performance
With the chain pattern, if we want to assign labels after ensuring all the elements are flex, we'd need to write something like this:
```
els
  .map(ensureFlex)
  .map(assignLabel)
  ...
```

That gives us twice as many iterations over the array of `els`.  So why not do everything in one iteration: 
```
els
  .map(el => { ensureFlex(el); assignLabel(el); }
  ...
```
The problem with this example is that elements are required further down the chain and there's no easy and pretty way (at least not that I'm aware of) to get around it rather than applying good ol' Functional Programming approach and HoF for that. There are also even more elegant ways to achieve it with FP approaches like [Transducers](https://medium.com/javascript-scene/transducers-efficient-data-processing-pipelines-in-javascript-7985330fe73d), but it's has a disadvantage of being overcomplicated.

#### Even more performance
Going further, we can get a little more by using `.forEach` instead of `.map` in this case:  
##### Cool:
- does not create a new array and as a result, saves time/space
##### Not cool:
- brings mutability to the stage, _which is not that terrible here, since we're modifying DOM objects anyway_

#### What else?
> tiny bit more succinct

That's a good thing too, right? =)

### Cons:
> when the functions are separate they can move around easier

That is sort of true, but all those functions were already explicitly dependent on `ensureFlex` method anyway.

> would/should update all other functional chains to use the same pattern (2 patterns instead of 1 until refactor)

That's what worries me the most. It's not easy to adopt such things on a project with such a huge codebase without a great amount of refactoring. And I believe this one worth switching the whole thing back to the chain pattern.

> was more straightforward before than with the abstraction

At first, it might take some time to wrap one's head around the FP approach, but often it leads to a more elegant way of solving the problems.


### Epilogue
Yeah, I'm trying to sell FP here, maybe unreasonable. But I'm more than fine to change it back to the old approach.
And thank you for your kind words. I'm still getting into open-source contribution, and such words from a person like you mean a lot! =)

Reply:
so, after lots of thinking, here's where I'm at:

**i like HoF's and FP concepts a lot**: 
that being said, they're a slippery slope here (i'd want to bring more and more FP concepts to the codebase), and i feel like we could get really wrapped up making performance updates. i totally agree, looping over each of the elements over and over immutably is memory overhead, and HoF (or really any refactor that just loops once) is a considerable gain. 

BUT, most the time it's 1 element in the collection. Rarely (i assume i'm one of the few power users) i make a selection that's 10+, and even then, the collection is small enough that the perf issues arent in the chain functions. essentially, there's other places to go focus our perf, and we should take a holistic look at making those corrections. let's introduce more FP complexity, fun and perf/efficiency later.

So!
if you could please, revert some of the code to the previous chain strategy. just so it matches the other tool tactics. for now. we'll keep this PR about adding labels and not about optimizations. 

- [x] all the tests pass we have (perhaps you want to write one that verifies a label is showing when using the align tool?)
- [x] all my smoke tests and manual QA passed great

I'm ready to merge this once the HoF code has been changed back! then, if you're down, we can chat about another flexbox ticket i think you'd do great at, or we can talk about 1 or 2 spots where there are real perf issues that are easily solvable but i havent had time to attack. 

thoughts!? i hope this isnt deterring, you've been a stellar contributor so far, and i'd like to keep working together. 

thanks!

Reply:
[made a ticket](https://github.com/GoogleChromeLabs/ProjectVisBug/issues/376) for that perf issue i allude to in my comment above 👍